crates/matrix-sdk-sled/src/state_store.rs

```diff
@@ -505,14 +498,41 @@ impl SledStore {
 
                     for (room, events) in &changes.stripped_members {
                         for event in events.values() {
+                            let key = (room, &event.state_key);
+
+                            match event.content.membership {
+                                MembershipState::Join => {
+                                    stripped_joined.insert(
+                                        self.encode_key(STRIPPED_JOINED_USER_ID, &key),
+                                        event.state_key.as_str(),
+                                    )?;
+                                    stripped_invited
+                                        .remove(self.encode_key(STRIPPED_INVITED_USER_ID, &key))?;
+                                }
+                                MembershipState::Invite => {
+                                    stripped_invited.insert(
+                                        self.encode_key(STRIPPED_INVITED_USER_ID, &key),
+                                        event.state_key.as_str(),
+                                    )?;
+                                    stripped_joined
+                                        .remove(self.encode_key(STRIPPED_JOINED_USER_ID, &key))?;
+                                }
+                                _ => {
+                                    stripped_joined
+                                        .remove(self.encode_key(STRIPPED_JOINED_USER_ID, &key))?;
+                                    stripped_invited
+                                        .remove(self.encode_key(STRIPPED_INVITED_USER_ID, &key))?;
+                                }
+                            }
                             stripped_members.insert(
-                                self.encode_key(
-                                    STRIPPED_ROOM_MEMBER,
-                                    (room, event.state_key.to_string()),
-                                ),
+                                self.encode_key(STRIPPED_ROOM_MEMBER, &key),
                                 self.serialize_event(&event)
                                     .map_err(ConflictableTransactionError::Abort)?,
                             )?;
+
+                            members.remove(self.encode_key(MEMBER, &key))?;
```

Comment:
This also just touches the members while it ignores other things.

Also I'm not sure about this. Our room becomes archived and we serve the state from the stripped buckets, but the spec mentions, or used to mention, this:

> If the client joins the room then the current state will be given as a delta against the archived state not the invite_state.

Now this is related to this issue: https://github.com/matrix-org/matrix-spec/issues/814

Synapse does return the whole state once you rejoin a left room so I don't think that this will break anything, though I would still just leave it be. It does protect us from e2ee downgrade attacks between rejoins.

Reply:
Hmm... this honestly becomes a lot of logic for a rather simple kv-store to manage, and very easy for an implementation to not adhere to proper. We might have a bunch of integration tests, but many of these aspects are not tested for, and the assumption that an implementer knows about this edge cases of the matrix spec sounds daunting to me. I think we should rethink the API and where the logic should lie for state stores going forward...

Reply:
I think we can just skipping the removal here, not removing the state her is fine:
1. If servers return the full state, it will just overwrite the old ones
2. If servers return the delta from the last time we were joined

So I don't think we need any additional logic here.

Reply:
you remove the current removing code? or leave it as is?

Reply:
I mean remove the current removing code.